src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/Target_ARM64/ARM64ReadyToRunGenericHelperNode.cs

```diff
@@ -15,30 +15,230 @@ partial class ReadyToRunGenericHelperNode
         }
 
         protected void EmitDictionaryLookup(NodeFactory factory, ref ARM64Emitter encoder, Register context, Register result, GenericLookupResult lookup, bool relocsOnly)
         {
-            throw new NotImplementedException();
+            // INVARIANT: must not trash context register
+
+            // Find the generic dictionary slot
+            int dictionarySlot = 0;
+            if (!relocsOnly)
+            {
+                // The concrete slot won't be known until we're emitting data - don't ask for it in relocsOnly.
+                dictionarySlot = factory.GenericDictionaryLayout(_dictionaryOwner).GetSlotForEntry(lookup);
+            }
+
+            // Load the generic dictionary cell
+            encoder.EmitLDR(result, context, dictionarySlot * factory.Target.PointerSize);
+
+            switch (lookup.LookupResultReferenceType(factory))
+            {
+                case GenericLookupResultReferenceType.Indirect:
+                    // Do another indirection
+                    encoder.EmitLDR(result, result);
+                    break;
+
+                case GenericLookupResultReferenceType.ConditionalIndirect:
+                    // Test result, 0x1
+                    // JEQ L1
+                    // mov result, [result-1]
+                    // L1:
+                    throw new NotImplementedException();
+
+                default:
+                    break;
+            }
         }
 
         protected sealed override void EmitCode(NodeFactory factory, ref ARM64Emitter encoder, bool relocsOnly)
         {
-            throw new NotImplementedException();
+            // First load the generic context into the context register.
+            EmitLoadGenericContext(factory, ref encoder, relocsOnly);
+
+            Register contextRegister = GetContextRegister(ref encoder);
+
+            switch (_id)
+            {
+                case ReadyToRunHelperId.GetNonGCStaticBase:
+                    {
+                        Debug.Assert(contextRegister == encoder.TargetRegister.Arg0);
+
+                        EmitDictionaryLookup(factory, ref encoder, encoder.TargetRegister.Arg0, encoder.TargetRegister.Result, _lookupSignature, relocsOnly);
+
+                        MetadataType target = (MetadataType)_target;
+                        if (!factory.PreinitializationManager.HasLazyStaticConstructor(target))
+                        {
+                            encoder.EmitRET();
+                        }
+                        else
+                        {
+                            // We need to trigger the cctor before returning the base. It is stored at the beginning of the non-GC statics region.
+                            encoder.EmitMOV(encoder.TargetRegister.Arg3, encoder.TargetRegister.Arg0);
+                            //encoder.EmitSUB(encoder.TargetRegister.Arg3, NonGCStaticsNode.GetClassConstructorContextStorageSize(factory.Target, target));
```

Comment:
I would expect GetClassConstructorContextStorageSize to be necessary.

Does this do the expected thing?

```csharp
class First<T> { public static int Number = new Random().Next(); }
class Second<T> { public static void Print() { Console.WriteLine(First<T>.Number); }

// Triggers the static constructor of First
Second<object>.Print();

// doesn't trigger it anymore
Second<object>.Print();
```

(The helper should be exercised within the `Second<T>.Print` body and trigger the static initializer of `First`.)

Reply:
Yes, it is needed. Thank you for spotting and the test case. 